From ee50f350478207405f7ab7ca816d6726eeedea44 Mon Sep 17 00:00:00 2001
From: abderraouf-b <boukhatem.abderraouf@gmail.com>
Date: Sat, 15 Mar 2025 11:07:29 +0100
Subject: [PATCH] libpcp: pcp_event_handler: check for integer overflow (#31)

* libpcp: pcp_event_handler: check for integer overflow

For systems with 32bit time_t, we can have an integer overflow calculating recv_lifetime.
Value could be truncated to LONG_MAX in this case.

* Check for overflow before assignement

---------

Co-authored-by: Abderraouf Boukhatem <abderraouf.boukhatem_ext@softathome.com>
---
 libpcp/src/pcp_event_handler.c | 8 +++++++-
 1 file changed, 7 insertions(+), 1 deletion(-)

diff --git a/libpcp/src/pcp_event_handler.c b/libpcp/src/pcp_event_handler.c
index 37f83dc..5a56f2a 100644
--- a/libpcp/src/pcp_event_handler.c
+++ b/libpcp/src/pcp_event_handler.c
@@ -37,6 +37,7 @@
 #include <errno.h>
 #include <time.h>
 #include <assert.h>
+#include <limits.h>
 
 #ifdef WIN32
 #include "pcp_win_defines.h"
@@ -480,7 +481,12 @@ static pcp_flow_event_e fhndl_received_success(pcp_flow_t *f,
     struct timeval ctv;
 
     PCP_LOG_BEGIN(PCP_LOGLVL_DEBUG);
-    f->recv_lifetime=msg->received_time + msg->recv_lifetime;
+    // avoid integer overflow
+    if (msg->recv_lifetime > (time_t) LONG_MAX - msg->received_time) {
+        f->recv_lifetime = LONG_MAX;
+    } else {
+        f->recv_lifetime = msg->received_time + msg->recv_lifetime;
+    }
     if ((f->kd.operation == PCP_OPCODE_MAP)
             || (f->kd.operation == PCP_OPCODE_PEER)) {
         f->map_peer.ext_ip=msg->assigned_ext_ip;
